library/alloc/src/collections/btree/map/entry.rs

```diff
@@ -312,22 +313,39 @@ impl<'a, K: Ord, V> VacantEntry<'a, K, V> {
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn insert(self, value: V) -> &'a mut V {
-        let out_ptr = match self.handle.insert_recursing(self.key, value) {
-            (Fit(_), val_ptr) => {
-                // SAFETY: We have consumed self.handle and the handle returned.
-                let map = unsafe { self.dormant_map.awaken() };
-                map.length += 1;
-                val_ptr
-            }
-            (Split(ins), val_ptr) => {
-                drop(ins.left);
+        let out_ptr = match self.handle {
+            None => {
                 // SAFETY: We have consumed self.handle and the reference returned.
                 let map = unsafe { self.dormant_map.awaken() };
-                let root = map.root.as_mut().unwrap();
-                root.push_internal_level().push(ins.kv.0, ins.kv.1, ins.right);
+                let (root, out_ptr) = {
+                    // Create a new leaf as root node and push the key-value pair into it.
+                    let mut root = NodeRef::new_leaf();
+                    let mut leaf_node = root.borrow_mut();
+                    leaf_node.push(self.key, value);
+                    let out_ptr: *mut V = leaf_node.last_kv().into_val_mut() as *mut V;
```

Comment:
The `*mut V` comes from `Handle::insert_fit`, then forwarded by `Handle::insert`, and then reborrowed by `Entry::insert`. For this special case (inserting into a new map), I agree with you that returning `&mut V` directly is better.

Reply:
For clarity, with "the first *mut V" I meant the type annotation, in addition to the "as" cast. I  was too lazy to look up the right terms (it's not type ascription or declaration); and I'm still too lazy to figure out "as" - it's a "primitive cast", [says the book](https://doc.rust-lang.org/book/appendix-01-keywords.html), but "cast" appears only twice elsewhere.

Reply:
I do have a physical book (O'Reilly's Programming Rust): it says the "as operator" converts, doesn't even mention casting.

Reply:
I messed up the `*mut V` you mentioned with other places and I thought you meant the conversion from `&mut V` to `*mut V` and then back to `&mut V`. So there has been another version discussed at https://github.com/rust-lang/rust/pull/92962#r820095892.

> For clarity, with "the first *mut V" I meant the type annotation, in addition to the "as" cast.

Then I realised you meant the two `*mut V` right here in the same statement.

> And the first *mut V seems pretty superfluous.

I do agree with you. Thanks for pointing out that.

Reply:
To be consistent with other functions like `insert` and `insert_fit`, I have made `push` to return a `*mut V` instead of `&mut V`, also because the `&mut V` is immediately coerced to a `*mut V` after returned.

Reply:
I think that's a bad idea. If it works with a high level `&mut V`, why resort to a lower level  `*mut V`? Okay, "it" is just the `push` function itself. But `MaybeUninit::write` returns a reference because it can, and so should `push`.

`insert` unfortunately can't return a `&mut V` because that would be a lie: it's not an exclusive reference since the function also returns an `InsertResult` that points to the same key-value pair. Come to think of it, why? Well, for no reason. I can yank out the handle wrapper of `InsertResult` with no problem. But let's keep that out of this PR, I'm starting to stall it here.

Reply:
To be clarify, I mean it is `insert_fit`which returns a `*mut V` instead of `&mut V` in the fact that it *can* returns a `&mut V`.
https://github.com/rust-lang/rust/blob/3d1eaf4b6203ab0168da5b99049942385aa753e0/library/alloc/src/collections/btree/node.rs#L845-L856

> insert unfortunately can't return a &mut V because that would be a lie: it's not an exclusive reference since the function also returns an InsertResult that points to the same key-value pair.

Sure, I agree with you about `Handle::insert`.

https://github.com/rust-lang/rust/blob/3d1eaf4b6203ab0168da5b99049942385aa753e0/library/alloc/src/collections/btree/node.rs#L623-L632
As you can see, `push` does what `insert_fit` does just like `Vec::push` does what `Vec::insert` does, except the former inserts at the end, and the latter inserts inside the node.

 I guess the reason why `insert_fit` returns a `*mut V` instead of `&mut V` is, that its caller `Handle::insert` (which returns an `InsertResult` with a `*mut V`) actually needs a `*mut V`, and even given a `&mut V`, the caller should also cast it to a `*mut V` because of the reason you have just mentioned. So does `push` with its caller `Entry::insert`.

Even though `Entry::insert` returns a `&mut V` in the end, it cannot preserve the returned `&mut V` since the temporarily created root node has to be moved into `map.root`. Our borrow checker doesn't allow that.

Reply:
I don't think I fully understand your point. I think `insert_fit` should ideally return a `&mut V`, but since it is a private function only used right below and always converted to `*mut _`, there's not much point in making it look nicer. But `push` is a public function (from the perspective of the private node module), it should be held to a higher standard.

Reply:
> But push is a public function

Now I get what you are concerned of. I didn't realise that `push` is public.

> it should be held to a higher standard.

Sure it is.

Reply:
I see now that it makes little sense to let `insert_fit` (on leaf nodes) return `&mut V `, because that doesn't live long enough. It cannot return `&'a mut V `, because the borrow checker won't allow that. It cannot consume `self` either.

Reply:
I don't think so. `assume_init_mut` borrows from `self.node.val_area_mut(self.idx)`, which borrows from `self.node`. Since `self` is passed by reference, it is no problem to return a `&mut V` which lives as long as `&mut self` (both of which lifetimes are eliminated).

Reply:
Sorry, I tried to clarify the first version of the comment and pasted into the wrong sentence. Yes, there's no problem letting it return `&mut V`, and I think that's slightly better, but even then, we need a `&'a mut V`.

Reply:
So I tried at https://github.com/ssomers/rust/tree/btree_prune_insert_2 to pass the `&'a mut V` directly back to the caller, and have one less case where the conversion to `*mut V` happens, but it's hardly worth the trouble. So instead I settled on the slight simplification #94699.

Reply:
```rust
            InsertResult::Fit(unsafe { &mut *(val_ref as *mut _) })
```
https://github.com/ssomers/rust/blob/160dd614ce752c9e7f447190273e2849a5b41867/library/alloc/src/collections/btree/node.rs#L867

Oh, I think it's even more tricky than the current version. I really thought it was totally safe until I see this `unsafe` block.  This just splits up a single `unsafe` block (at the end of `Entry::insert`) to two, and one of which is hidden inside (in the `node` module).

Anyway, #94699 looks much better.